### PR TITLE
nerf rockets

### LINKED ISSUE #562: Slow down player when wielding rocket launcher
-20% when rocket launcher is equipped

### LINKED ISSUE #563: Rate limit rockets

### PR DESCRIPTION
**Pull request by olliy1or**
Closes https://github.com/MT-CTF/capturetheflag/issues/562 and Closes https://github.com/MT-CTF/capturetheflag/issues/563
and maybe this too? https://github.com/MT-CTF/capturetheflag/issues/569

### CODE REVIEW

**Comment by olliy1or** [comment]
this could close https://github.com/MT-CTF/capturetheflag/issues/562 ??

File: shooter_rocket/init.lua [hunk 1]
```diff
@@ -135,9 +135,26 @@ if shooter.config.enable_crafting == true then
 	})
 end
 
+local timer = 0
+minetest.register_globalstep(function(dtime)
+    timer = timer + dtime
+    if timer >= 0.1 then
```
Comment (LoneWolfHT) [inline shooter_rocket/init.lua]: The interval could be `1` or something similar. A second-long delay (at most in this case) isn't much. And we can just ignore the cases where the player loads and fires immediately after

File: shooter_rocket/init.lua [hunk 2]
```diff
@@ -135,9 +135,26 @@ if shooter.config.enable_crafting == true then
...
+        for _, player in ipairs(minetest.get_connected_players()) do
+            local name = player and player:get_player_name()
+            local wielditem = player:get_wielded_item()
+            if wielditem:get_name() == "shooter_rocket:rocket_gun_loaded" then
+                local wielditem = player:get_wielded_item()
+                player:set_physics_override({
+                    speed = 0.7
+                })
+                timer = 0
```
Comment (LoneWolfHT) [inline shooter_rocket/init.lua]: `timer = 0` should be moved beneath `if timer >= 0.1 then`

File: shooter_rocket/init.lua [hunk 3]
```diff
@@ -135,9 +135,26 @@ if shooter.config.enable_crafting == true then
...
+    timer = timer + dtime
+    if timer >= 0.1 then
+        for _, player in ipairs(minetest.get_connected_players()) do
+            local name = player and player:get_player_name()
+            local wielditem = player:get_wielded_item()
+            if wielditem:get_name() == "shooter_rocket:rocket_gun_loaded" then
+                local wielditem = player:get_wielded_item()
+                player:set_physics_override({
+                    speed = 0.7
+                })
```
Comment (LoneWolfHT) [inline shooter_rocket/init.lua]: Might want to use the physics mod for this, if it can be found by shooter. Otherwise it might conflict with other mods

**Review by LoneWolfHT — CHANGES_REQUESTED** [review]

File: shooter_rocket/init.lua (hunk 1 above)
Comment (olliy1or) [reply shooter_rocket/init.lua]: done.

File: shooter_rocket/init.lua (hunk 2 above)
Comment (olliy1or) [reply shooter_rocket/init.lua]: done.

File: shooter_rocket/init.lua (hunk 3 above)
Comment (olliy1or) [reply shooter_rocket/init.lua]: And done.

File: shooter_rocket/init.lua [hunk 4]
```diff
@@ -17,6 +17,17 @@ with this program; if not, write to the Free Software Foundation, Inc.,
 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 ]]--
 
+local plcooldown = {}
+local COOLDOWN = 5
+
+minetest.register_on_joinplayer(function(player)
+	plcooldown[player:get_player_name()] = 0
```
Comment (Panquesito7) [inline shooter_rocket/init.lua]: Maybe the cooldown should be saved in the player's meta or something like that. I could simply leave, join, take the rocket, and shoot again.

File: shooter_rocket/init.lua [hunk 5]
```diff
@@ -70,33 +81,43 @@ minetest.register_tool("shooter_rocket:rocket_gun_loaded", {
...
-		itemstack = "shooter_rocket:rocket_gun 1 "..itemstack:get_wear()
-		if pointed_thing.type ~= "nothing" then
-			local pointed = minetest.get_pointed_thing_position(pointed_thing)
-			if vector.distance(user:get_pos(), pointed) < 8 then
-				shooter.blast(pointed, 2, 50, 7)
-				return itemstack
+		if plcooldown[user:get_player_name()] ~= 0 then
+			minetest.chat_send_player(user:get_player_name(), "Your rocket has a cooldown!")
+		else
+			if not minetest.setting_getbool("creative_mode") then
```
Comment (Panquesito7) [inline shooter_rocket/init.lua:87]: The method you're using is deprecated. Use this one instead.
```suggestion
			if not minetest.settings:get_bool("creative_mode") then
```

File: shooter_rocket/init.lua (hunk 4 above)
Comment (olliy1or) [reply shooter_rocket/init.lua]: Don't think people would leave and join just so they can shoot a rocket, also by the time they could leave and join the timer would most likely be up

File: shooter_rocket/init.lua [hunk 6]
```diff
@@ -70,33 +81,47 @@ minetest.register_tool("shooter_rocket:rocket_gun_loaded", {
...
-			itemstack:add_wear(65535 / 50)
-		end
-		itemstack = "shooter_rocket:rocket_gun 1 "..itemstack:get_wear()
-		if pointed_thing.type ~= "nothing" then
-			local pointed = minetest.get_pointed_thing_position(pointed_thing)
-			if vector.distance(user:get_pos(), pointed) < 8 then
-				shooter.blast(pointed, 2, 50, 7)
-				return itemstack
+		if plcooldown[user:get_player_name()] ~= 0 then
+		hud_event.new(user:get_player_name(), {
```
Comment (Panquesito7) [inline shooter_rocket/init.lua:81]: `hud_event` should be added in `shooter_rocket`'s dependencies.

**Review by Panquesito7 — CHANGES_REQUESTED** [review]

File: shooter_rocket/init.lua (hunk 4 above)
Comment (olliy1or) [reply shooter_rocket/init.lua]: done think ppl will rejoin to bypass a 5sec cooldown

**Review by Panquesito7 — APPROVED** [review]

File: shooter_rocket/init.lua [hunk 7]
```diff
@@ -17,6 +17,17 @@ with this program; if not, write to the Free Software Foundation, Inc.,
 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 ]]--
 
+local plcooldown = {}
+local COOLDOWN = 5
+
+minetest.register_on_joinplayer(function(player)
+	plcooldown[player:get_player_name()] = 0
+end)
```
Comment (LoneWolfHT) [inline shooter_rocket/init.lua]:
```suggestion
```

Could remove this and apply my other suggestions

File: shooter_rocket/init.lua [hunk 8]
```diff
@@ -70,33 +81,47 @@ minetest.register_tool("shooter_rocket:rocket_gun_loaded", {
...
-		if not minetest.setting_getbool("creative_mode") then
-			itemstack:add_wear(65535 / 50)
-		end
-		itemstack = "shooter_rocket:rocket_gun 1 "..itemstack:get_wear()
-		if pointed_thing.type ~= "nothing" then
-			local pointed = minetest.get_pointed_thing_position(pointed_thing)
-			if vector.distance(user:get_pos(), pointed) < 8 then
-				shooter.blast(pointed, 2, 50, 7)
-				return itemstack
+		if plcooldown[user:get_player_name()] ~= 0 then
```
Comment (LoneWolfHT) [inline shooter_rocket/init.lua]:
```lua
if plcooldown[user:get_player_name()] then
```

File: shooter_rocket/init.lua [hunk 9]
```diff
@@ -70,33 +81,47 @@ minetest.register_tool("shooter_rocket:rocket_gun_loaded", {
...
+			if pointed_thing.type ~= "nothing" then
+				local pointed = minetest.get_pointed_thing_position(pointed_thing)
+				if vector.distance(user:get_pos(), pointed) < 8 then
+					shooter.blast(pointed, 2, 50, 7)
+					return itemstack
 				end
 			end
+			local pos = user:get_pos()
+			local dir = user:get_look_dir()
+			local yaw = user:get_look_horizontal()
```
Comment (LoneWolfHT) [inline shooter_rocket/init.lua:100]: Should add
```lua
local username = user:get_player_name()
```

You could then change all the following code that calls user:get_player_name()

File: shooter_rocket/init.lua [hunk 10]
```diff
@@ -70,33 +81,47 @@ minetest.register_tool("shooter_rocket:rocket_gun_loaded", {
...
+					local ent = obj:get_luaentity()
+					if ent then
+						ent.user = user:get_player_name()
+					end
+				end
+			end
+			plcooldown[user:get_player_name()] = COOLDOWN
+			minetest.after(COOLDOWN, function(user)
+				if plcooldown[user:get_player_name()] then
+					plcooldown[user:get_player_name()] = 0
```
Comment (LoneWolfHT) [inline shooter_rocket/init.lua]:
```suggestion
					plcooldown[user:get_player_name()] = nil
```
If you implement the above suggestions

File: shooter_rocket/init.lua (hunk 8 above)
Comment (olliy1or) [reply shooter_rocket/init.lua]: Done :heavy_check_mark:

File: shooter_rocket/init.lua (hunk 9 above)
Comment (olliy1or) [reply shooter_rocket/init.lua:100]: Done :heavy_check_mark:

**Pull request closed without merging**